packages/flutter/lib/src/painting/image_stream.dart

```diff
@@ -2,35 +2,140 @@
 // Use of this source code is governed by a BSD-style license that can be
 // found in the LICENSE file.
 
-
 import 'dart:async';
 import 'dart:ui' as ui show Image, Codec, FrameInfo;
 import 'dart:ui' show hashValues;
 
 import 'package:flutter/foundation.dart';
 import 'package:flutter/scheduler.dart';
 
+/// A reference to a [ui.Image] object.
+///
+/// To get an ImageHandle, clients should call [ImageInfo.obtainHandle].
+///
+/// Clients of an [ImageStream] should call [dispose] on this object when they
+/// will no longer need to use the [ImageInfo.image] vended to them in
+/// [ImageStreamCompleter.setImage]. When all clients of the stream have
+/// disposed their handles, the underlying [ui.Image] will be disposed. Clients
+/// must not call dispose more than once on the reference they receive from
+/// [ImageInfo.obtainHandle].
+///
+/// Clients should _not_ call [ui.Image.dispose] directly, as it will prevent
+/// other clients from using the same image reference should they need to.
+class ImageHandle {
+  /// Do not use this, use [_create] instead to make sure that ImageHandles
+  /// for the same [ui.Image] are the same.
+  ImageHandle._(this._image)
+    : assert(_image != null);
+
+  factory ImageHandle._create(ui.Image image) {
+    assert(image != null);
+    _handles[image] ??= ImageHandle._(image);
+    return _handles[image]!;
+  }
+
+  static final Expando<ImageHandle> _handles = Expando<ImageHandle>();
+
+  final ui.Image _image;
+  int _refCount = 0;
+
+  /// The number of references to this [ImageHandle].
+  ///
+  /// This always returns null if asserts are disabled, and is only meant for
+  /// use in debug mode or tests.
+  int? get debugRefCount {
```

Comment:
generally we only add debug only code that is actually used by the framework (in debug mode), and not just for testing. Seems like a lot to pollute the API just for a unit test

Reply:
We definitely have other cases where it's added only for tests (for example, debugAssertNoTransientCallbacks on SchedulerBinding). I can mark it @visibleForTesting if you think that's better. 

This could be helpful for application code, if someone is developing a new type of image provider and can't figure out why they still have references to it and who might be adding them.  Alternatively, the tests that use it could instead just disable the image cache to avoid it adding more refs.

Reply:
I think in this case it feels sort of me to me, because the API is otherwise so small. I admit though, this isn't really that important, but if you think you can test the ImageHandle without the getters, it would be better to YAGNI, right?